tests/integration/high_availability/test_async_replication_upgrade.py

```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import time
+from collections.abc import Generator
+
+import jubilant
+import pytest
+from jubilant import Juju
+from tenacity import Retrying, stop_after_attempt
+
+from .. import architecture
+from .high_availability_helpers_new import (
+    get_app_leader,
+    get_app_units,
+    get_db_max_written_value,
+    wait_for_apps_status,
+)
+
+DB_APP_NAME = "postgresql"
+DB_APP_1 = "db1"
+DB_APP_2 = "db2"
+DB_TEST_APP_NAME = "postgresql-test-app"
+
+MINUTE_SECS = 60
+
+logging.getLogger("jubilant.wait").setLevel(logging.WARNING)
+
+
+@pytest.fixture(scope="module")
+def first_model(juju: Juju, request: pytest.FixtureRequest) -> Generator:
+    """Creates and return the first model."""
+    yield juju.model
+
+
+@pytest.fixture(scope="module")
+def second_model(juju: Juju, request: pytest.FixtureRequest) -> Generator:
+    """Creates and returns the second model."""
+    model_name = f"{juju.model}-other"
+
+    logging.info(f"Creating model: {model_name}")
+    juju.add_model(model_name)
+
+    yield model_name
+    if request.config.getoption("--keep-models"):
+        return
+
+    logging.info(f"Destroying model: {model_name}")
+    juju.destroy_model(model_name, destroy_storage=True, force=True)
+
+
+@pytest.fixture()
+def first_model_continuous_writes(first_model: str) -> Generator:
+    """Starts continuous writes to the cluster for a test and clear the writes at the end."""
+    model_1 = Juju(model=first_model)
+    application_unit = get_app_leader(model_1, DB_TEST_APP_NAME)
+
+    logging.info("Clearing continuous writes")
+    model_1.run(
+        unit=application_unit, action="clear-continuous-writes", wait=120
+    ).raise_on_failure()
+
+    logging.info("Starting continuous writes")
+
+    for attempt in Retrying(stop=stop_after_attempt(10), reraise=True):
+        with attempt:
+            result = model_1.run(unit=application_unit, action="start-continuous-writes")
+            result.raise_on_failure()
+
+            assert result.results["result"] == "True"
+
+    yield
+
+    logging.info("Clearing continuous writes")
+    model_1.run(
+        unit=application_unit, action="clear-continuous-writes", wait=120
+    ).raise_on_failure()
+
+
+def test_deploy(first_model: str, second_model: str, charm: str) -> None:
+    """Simple test to ensure that the MySQL application charms get deployed."""
+    configuration = {"profile": "testing"}
+    constraints = {"arch": architecture.architecture}
+
+    # TODO Deploy from edge
+    logging.info("Deploying postgresql clusters")
+    model_1 = Juju(model=first_model)
+    model_1.deploy(
+        charm=charm,
+        app=DB_APP_1,
+        base="ubuntu@24.04",
+        config=configuration,
+        constraints=constraints,
+        num_units=3,
+    )
+    model_2 = Juju(model=second_model)
+    model_2.deploy(
+        charm=charm,
```

Comment:
Pre-refresh check will be ran on the old code, so it will fail on edge (`leader` vs. `standby leader`).